Sort "What's New" page chronologically (latest-oldest)

timfbfbfbfb opened the pull request:
### Features:
1. Sort `What's New` page chronologically, in latest-oldest order
2. Fetch recent videos from subscribed channels in multiple background threads parallel, this speeds up the download process quite a lot
3. Incrementally insert video items to `What's New` list according to the publish time, so users could see the first meaning paint earlier

### Limitations:
1. Publish time is retrieved from Youtube RSS feeds. This would cost users extra mobile data, though the RSS feed XML files are quite small compared to the Youtube channel video HTML pages (each RSS feed XML file is approximately 1/5 to 1/4 file size of the HTML page)
2. Since Youtube RSS feed provides at most the latest 15 videos from the channel, the `What's New` page only shows the latest 15 videos from each subscribed channel

### Remarks:
I missed the PR TeamNewPipe/NewPipe#2309 at first and had just noticed it implemented the similar chronological sort feature in the middle of the development, but I still decided to create this PR for two reasons.

1. I found that this PR is quite different from TeamNewPipe/NewPipe#2309, for instance feature 3 has not been implemented in TeamNewPipe/NewPipe#2309. I have downloaded the apk of TeamNewPipe/NewPipe#2309 and tried it out, the channel grouping feature is really cool, I think this PR and TeamNewPipe/NewPipe#2309 could complement each other well.
2. Many NewPipe users want the chronological sort feature so badly. This PR is relatively small, maybe it would be easier for other contributors to review and merge

### APK for reviewers to test:
Self-signed release APK
[app-afa9fa55.apk.zip](https://github.com/TeamNewPipe/NewPipe/files/4271494/app-afa9fa55.apk.zip)

**Since this APK is self-signed, `Play Protect` may show a warning with that, just select "Install anyway" to install**

### Related issues
Close TeamNewPipe/NewPipe#739
Close TeamNewPipe/NewPipe#822


### Related PRs
Merge together with TeamNewPipe/NewPipeExtractor#189


- [x] I carefully read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md) and agree to them.

snappyapple632 commented:
I'm seeing duplication on certain videos.
![Capture _2019-08-23-06-13-07](https://user-images.githubusercontent.com/40573129/63595714-74e8d080-c56e-11e9-9bad-10c4558595cc.png)
![Capture _2019-08-23-06-13-19](https://user-images.githubusercontent.com/40573129/63595716-75816700-c56e-11e9-85fe-36e6486fa5ec.png)
![Capture _2019-08-23-06-13-33](https://user-images.githubusercontent.com/40573129/63595717-75816700-c56e-11e9-85f1-f60bce078c49.png)

timfbfbfbfb commented:
@snappyapple632  Thanks for your feedback!
I think I just accidentally call `startLoading()` twice when the app is first started. I did not even notice the duplicated videos because I tested the app with too many subscribed channels.
Anyway, I have fixed that and tested on my side, and I updated the APK in the description. Could you please help to verify that?

snappyapple632 commented:
The issue has been solved.

chilliger commented:
Tested the new apk. No duplication for me. But every time loading the whats new feed I get the following exception. This is also happening when I switch back to Newpipe from an other app

`{
   "user_action": "subscription",
   "request": "https:\/\/www.youtube.com\/channel\/UC6pfOVz7GxTei7hEMlm7cCA",
   "content_language": "DE",
   "service": "YouTube",
   "package": "org.schabi.newpipe.pr2547",
   "version": "0.17.1.pr2547",
   "os": "Linux Android 9 - 28",
   "time": "2019-08-23 20:09",
   "exceptions": [
      "org.schabi.newpipe.extractor.exceptions.ParsingException: Could not get channel name\n\tat org.schabi.newpipe.extractor.services.youtube.extractors.YoutubeChannelExtractor.getName(YoutubeChannelExtractor.java:126)\n\tat org.schabi.newpipe.extractor.channel.ChannelInfo.getInfo(ChannelInfo.java:62)\n\tat org.schabi.newpipe.extractor.channel.ChannelInfo.getInfo(ChannelInfo.java:49)\n\tat org.schabi.newpipe.util.ExtractorHelper.lambda$getChannelInfo$4(ExtractorHelper.java:123)\n\tat org.schabi.newpipe.util.-$$Lambda$ExtractorHelper$u5W7VszTe8AoEexIsFM9huQfbkM.call(Unknown Source:4)\n\tat io.reactivex.internal.operators.single.SingleFromCallable.subscribeActual(SingleFromCallable.java:44)\n\tat io.reactivex.Single.subscribe(Single.java:3438)\n\tat io.reactivex.internal.operators.single.SingleDoOnSuccess.subscribeActual(SingleDoOnSuccess.java:35)\n\tat io.reactivex.Single.subscribe(Single.java:3438)\n\tat io.reactivex.internal.operators.maybe.MaybeFromSingle.subscribeActual(MaybeFromSingle.java:41)\n\tat io.reactivex.Maybe.subscribe(Maybe.java:4154)\n\tat io.reactivex.internal.operators.maybe.MaybeConcatArray$ConcatMaybeObserver.drain(MaybeConcatArray.java:153)\n\tat io.reactivex.internal.operators.maybe.MaybeConcatArray$ConcatMaybeObserver.request(MaybeConcatArray.java:78)\n\tat io.reactivex.internal.operators.flowable.FlowableElementAtMaybe$ElementAtSubscriber.onSubscribe(FlowableElementAtMaybe.java:66)\n\tat io.reactivex.internal.operators.maybe.MaybeConcatArray.subscribeActual(MaybeConcatArray.java:42)\n\tat io.reactivex.Flowable.subscribe(Flowable.java:14479)\n\tat io.reactivex.internal.operators.flowable.FlowableElementAtMaybe.subscribeActual(FlowableElementAtMaybe.java:36)\n\tat io.reactivex.Maybe.subscribe(Maybe.java:4154)\n\tat io.reactivex.internal.operators.maybe.MaybeToSingle.subscribeActual(MaybeToSingle.java:46)\n\tat io.reactivex.Single.subscribe(Single.java:3438)\n\tat io.reactivex.internal.operators.maybe.MaybeFromSingle.subscribeActual(MaybeFromSingle.java:41)\n\tat io.reactivex.Maybe.subscribe(Maybe.java:4154)\n\tat io.reactivex.internal.operators.maybe.MaybeSubscribeOn$SubscribeTask.run(MaybeSubscribeOn.java:54)\n\tat io.reactivex.internal.schedulers.ScheduledDirectTask.call(ScheduledDirectTask.java:38)\n\tat io.reactivex.internal.schedulers.ScheduledDirectTask.call(ScheduledDirectTask.java:26)\n\tat java.util.concurrent.FutureTask.run(FutureTask.java:266)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)\n\tat java.lang.Thread.run(Thread.java:764)\nCaused by: java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String org.jsoup.nodes.Element.attr(java.lang.String)' on a null object reference\n\tat org.schabi.newpipe.extractor.services.youtube.extractors.YoutubeChannelExtractor.getName(YoutubeChannelExtractor.java:124)\n\t... 28 more\n"
   ],
   "user_comment": ""
}`

I don't know if this is related to this change.

chilliger commented:
Another note on this. Is it possible to cache the order in some way to improve the loading time and to minimize the jumping effect when new found videos get inserted. Maybe setting up the background fetching with how often it should be done could be an option?

snappyapple632 commented:
Use the Markdown converter next time @chilliger

timfbfbfbfb commented:
> Is it possible to cache the order in some way to improve the loading time and to minimize the jumping effect when new found videos get inserted. 

I think finding the correct positions of the video items would not contribute too much to the loading time and the jumping effect.
Like I explained in the description, I fetch video items from all subscribed channels in the background threads and insert the video items to the list incrementally in the main UI thread, a little bit of jumping effect is unavoidable. I think this is a trade-off. I could avoid jumping effect if I wait for all HTTP responses before displaying the list, however, the loading time would be very long.

> Maybe setting up the background fetching with how often it should be done could be an option?

The current implementation would just fetch the videos from the subscribed channels once.

coolStudent123 commented:
@timfbfbfbfb can you add a notification with progress so that the user knows when all subscriptions have been checked ? This was implemented in https://github.com/TeamNewPipe/NewPipe/pull/2309 . In the current implementation it is unclear when the subscription check is done and while I like already seeing the videos and being able to scroll already, I have to scroll back up constantly since I never really know when all subscriptions have been checked. IMO this could be added later and I just want to see this merged asap but that's just my opinion.

timfbfbfbfb commented:
@coolStudent123 That is a good suggestion, I think we can add a horizontal progress bar at the very top to indicate the loading progress. But as you have said, I want to keep this PR small enough to merge, maybe we can create a separate issue for that. I can also help working on that later.

opusforlife2 commented:
Also closes #822 .

rakshazi commented:
Hello, any update on this?
No news for last 2 weeks :(

esalgado commented:
I tested the build for several days and runs very well without issues for me.

rakshazi commented:
So. may be it's time to merge that PR?

Stypox commented on app/src/main/java/org/schabi/newpipe/local/feed/FeedFragment.java:
```diff
@@ -159,220 +193,173 @@ public void readFrom(@NonNull Queue<Object> savedObjects) throws Exception {
                 .onErrorReturnItem(Collections.emptyList())
-                .observeOn(AndroidSchedulers.mainThread())
-                .subscribe(this::handleResult, this::onError);
+                .observeOn(Schedulers.io())
```
Are you sure this is ok? `handleResult` and `onError` would be run on a thread different from the main one, and since they do gui stuff this could potentially create problems. I could be wrong, though ;-)

timfbfbfbfb replied:
Since all the UI related operations are executed inside the callbacks of `Handler::post`, they will eventually be executed in the main thread. I think this is okay.

Stypox replied:
Oh, ok. Thank you for the clarification ;-)

Stypox commented on app/src/main/java/org/schabi/newpipe/local/feed/FeedFragment.java:
```diff
@@ -159,220 +193,173 @@ public void readFrom(@NonNull Queue<Object> savedObjects) throws Exception {
+        numChannels.set(filteredResult.size());
+        compositeDisposable.add(
+            Flowable.fromIterable(filteredResult)
+                .observeOn(Schedulers.io())
```
Here, too.

Stypox commented on app/src/main/java/org/schabi/newpipe/local/feed/FeedFragment.java:
```diff
@@ -159,220 +193,173 @@ public void readFrom(@NonNull Queue<Object> savedObjects) throws Exception {
+        compositeDisposable.add(
+            subscriptionService.getChannelInfo(subscriptionEntity)
+                .observeOn(Schedulers.io())
+                .onErrorComplete((@io.reactivex.annotations.NonNull Throwable throwable) -> FeedFragment.super.onError(
```
Can't `@io.reactivex.annotations.NonNull` be imported and referenced just as `@NonNull`?

Stypox commented on app/src/main/java/org/schabi/newpipe/local/feed/FeedFragment.java:
```diff
@@ -417,6 +389,30 @@ private int howManyItemsToLoad() {
         return Math.max(MIN_ITEMS_INITIAL_LOAD, items);
     }
 
+    private String tryRetrieveVideoIdFromUrl(String url) {
```
Add `@Nullable` annotation, since you `return null`.

Stypox commented on app/src/main/java/org/schabi/newpipe/local/feed/FeedFragment.java:
```diff
@@ -417,6 +389,30 @@ private int howManyItemsToLoad() {
+        } catch (Exception ex) {
+            if (DEBUG)
+                Log.d(TAG, "Failed to retrieve video ID from \"" + url + "\"", ex);
+            return null;
```
Here we could probably show a Toast to the user, similarly to the one that's shown when a search result can't be parsed. It should contain the message "Sorry, some errors occoured" (the resource already exists in `strings.xml`) with a `REPORT` button.

Stypox commented on app/src/main/java/org/schabi/newpipe/util/ListUtils.java:
```diff
@@ -0,0 +1,23 @@
+import java.util.List;
+
+public class ListUtils {
+    public static <T, P> int binarySearchUpperBound(List<T> ls, P target, Function<T, P> getter, Comparator<P> cmp) {
```
Since this is a util method, you could add a Javadoc explaining inputs, return values, and optionally point out the complexity of the algorithm `O(log(n))`.

Stypox commented:
Sorry for the false positive, after further testing I found out that the "What's New" tab will not contain videos from non-YouTube channels. For example, try subscribing to a SoundCloud channel, then open "What's New". A crash report similar to this will be generated:

## Exception
* __User Action:__ something
* __Request:__ 
* __Content Language:__ GB
* __Service:__ none
* __Version:__ 0.17.3
* __OS:__ Linux Android 7.0 - 24


<details><summary><b>Crash log</b></summary><p>

```
org.schabi.newpipe.extractor.exceptions.ParsingException: The url is not a Youtube-URL
	at org.schabi.newpipe.extractor.services.youtube.linkHandler.YoutubeStreamLinkHandlerFactory.getId(YoutubeStreamLinkHandlerFactory.java:95)
	at org.schabi.newpipe.local.feed.FeedFragment.tryRetrieveVideoIdFromUrl(FeedFragment.java:395)
	at org.schabi.newpipe.local.feed.FeedFragment.lambda$getReceiveChannelInfoHandler$2(FeedFragment.java:254)
	at org.schabi.newpipe.local.feed.-$$Lambda$FeedFragment$Fx4XrJdA0lfiOjJgz-Oy8VCrWb0.accept(lambda)
	at io.reactivex.internal.operators.maybe.MaybeCallbackObserver.onSuccess(MaybeCallbackObserver.java:71)
	at io.reactivex.internal.operators.maybe.MaybeOnErrorComplete$OnErrorCompleteMaybeObserver.onSuccess(MaybeOnErrorComplete.java:67)
	at io.reactivex.internal.operators.maybe.MaybeObserveOn$ObserveOnMaybeObserver.run(MaybeObserveOn.java:103)
	at io.reactivex.Scheduler$DisposeTask.run(Scheduler.java:578)
	at io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:66)
	at io.reactivex.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:57)
	at java.util.concurrent.FutureTask.run(FutureTask.java:237)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:272)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
	at java.lang.Thread.run(Thread.java:776)

```
</p></details>
<hr>

Also, could you change "something" in the User Action field to something like "loading what's new" @timfbfbfbfb?

Stypox commented on app/src/main/java/org/schabi/newpipe/local/feed/FeedFragment.java:
```diff
@@ -417,6 +389,38 @@ private int howManyItemsToLoad() {
+            showSnackBarError(ex,
+                UserAction.SOMETHING_ELSE,
+                "none",
+                "",
```
Here you should put `url`, so that the chash report contains the URL in the field "Request: " and it is simpler to debug

timfbfbfbfb commented:
> Sorry for the false positive, after further testing I found out that the "What's New" tab will not contain videos from non-YouTube channels. For example, try subscribing to a SoundCloud channel, then open "What's New". A crash report similar to this will be generated:
> ## Exception
> 
>     * **User Action:** something
> 
>     * **Request:**
> 
>     * **Content Language:** GB
> 
>     * **Service:** none
> 
>     * **Version:** 0.17.3
> 
>     * **OS:** Linux Android 7.0 - 24
> 
> 
> Crash log

I updated the APK file in the description, SoundCloud channels should also be handled with the latest commit

Stypox commented:
What about MediaCCC? Did you also fix that?

Thanks :-D

timfbfbfbfb commented:
I did not see any MediaCCC channel extractor in NewPipeExtractor code base, probably there is no channels in MediaCCC. I am not familiar with MediaCCC, correct me if I am wrong.

Stypox commented:
Since channels in MediaCCC are called "conferences" you should look in [MediaCCCConferenceExtractor](https://github.com/TeamNewPipe/NewPipeExtractor/blob/dev/extractor/src/main/java/org/schabi/newpipe/extractor/services/media_ccc/extractors/MediaCCCConferenceExtractor.java)

AnXh3L0 commented:
Just tested the latest debug apk, works as intended, although I felt a minor slowdown but shouldn't be a big deal.

PeterHindes commented:
Hey, can you fix the conflicting files real quick? This is a great addition, cant wait to see it!

PeterHindes commented:
After some surface level testing of the debug apk provided in @Stypox comment, it seems to use a surprising amount of cpu while parsing. Is there a way to only fetch the most recent 3 videos? Maybe a menu allowing users to chose how many videos per Chanel they want to see. Would that eliminate some of the cpu stress?

n-xlkt commented:
@PeterHindes this is feature creep. i'd instead propose folks testing the debug apk post cpu usage while parsing so this isn't just n of 1.

timfbfbfbfb commented:
> After some surface level testing of the debug apk provided in @Stypox comment, it seems to use a surprising amount of cpu while parsing. Is there a way to only fetch the most recent 3 videos? Maybe a menu allowing users to chose how many videos per Chanel they want to see. Would that eliminate some of the cpu stress?

Since we are crawling the youtube video webpages, we cannot control the page size we fetch. So, if I remember correctly, we have to fetch at least 31 video info from each channel.

BTW, I didn't know that the bot would remove the merge approval after fixing the merge conflict, so maybe @Stypox has to review again :-P . So, should I fix the merge conflict myself next time or just let it be?

PeterHindes commented:
> > After some surface level testing of the debug apk provided in @Stypox comment, it seems to use a surprising amount of cpu while parsing. Is there a way to only fetch the most recent 3 videos? Maybe a menu allowing users to chose how many videos per Chanel they want to see. Would that eliminate some of the cpu stress?
> 
> Since we are crawling the youtube video webpages, we cannot control the page size we fetch. So, if I remember correctly, we have to fetch at least 31 video info from each channel.
> 
> BTW, I didn't know that the bot would remove the merge approval after fixing the merge conflict, so maybe @Stypox has to review again :-P . So, should I fix the merge conflict myself next time or just let it be?

Ok, thanks for considering it. It might be useful to have a favorites section of channels which are fetched first. Maybe there is also an extra option that could be passed which fetches a specific amount of videos. Also just for clarification how did the original "Whats New" section work? Was it just a random selection?

ZeDeedIsDone commented:
Is it just me, seems like no one is taking about it. I get errors if I try to view a channel, thus I can't even subscribe to a channel.

timfbfbfbfb commented:
> Is it just me, seems like no one is taking about it. I get errors if I try to view a channel, thus I can't even subscribe to a channel.
> Crash log

@ZeDeedIsDone 
Thanks for reporting the bug. 

I guess the error is caused by `new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSXXX")` in my NewPipeExtractor code. Since the 'XXX' timezone pattern is only supported in API level 24+, the app would crash on your Android 6 (API level 23) device.

I have used another workaround to parse the date string in NewPipeExtractor, it should fix the issue. Since I do not have Android 6 device for testing, maybe you can help to verify if the issue is fixed. The apk file in the description has been updated.

ZeDeedIsDone commented:
> > Is it just me, seems like no one is taking about it. I get errors if I try to view a channel, thus I can't even subscribe to a channel.
> > Crash log
> 
> @ZeDeedIsDone
> Thanks for reporting the bug.
> 
> I guess the error is caused by `new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSXXX")` in my NewPipeExtractor code. Since the 'XXX' timezone pattern is only supported in API level 24+, the app would crash on your Android 6 (API level 23) device.
> 
> I have used another workaround to parse the date string in NewPipeExtractor, it should fix the issue. Since I do not have Android 6 device for testing, maybe you can help to verify if the issue is fixed. The apk file in the description has been updated.

Works fine now, thanks! I did get random getUrlFeed errors for random channels in the what's new page but seems like after using the app for a few minutes, it looks like it's gone and loads the subscriptions fully. I'll edit this comment when the error does itself again.

vitSkalicky commented:
I've been using this PR for about two months and it works flawlessly. I would maybe change the loading spinner for a 'swipe to refesh' because now you cannot maually refresh and you are not sure if the loading has finished.

Ozyc commented:
I just got a crash whilst opening the app

## Exception
* __User Action:__ ui error
* __Request:__ App crash, UI failure
* __Content Language:__ FR
* __Service:__ none
* __Version:__ 0.17.4.version_4881970fe
* __OS:__ Linux Android 8.1.0 - 27


<details><summary><b>Crash log</b></summary><p>

```
java.lang.IllegalStateException: ViewHolder views must not be attached when created. Ensure that you are not passing 'true' to the attachToRoot parameter of LayoutInflater.inflate(..., boolean attachToRoot)
	at androidx.recyclerview.widget.RecyclerView$Adapter.createViewHolder(RecyclerView.java:6796)
	at androidx.recyclerview.widget.RecyclerView$Recycler.tryGetViewHolderForPositionByDeadline(RecyclerView.java:5975)
	at androidx.recyclerview.widget.RecyclerView$Recycler.getViewForPosition(RecyclerView.java:5858)
	at androidx.recyclerview.widget.RecyclerView$Recycler.getViewForPosition(RecyclerView.java:5854)
	at androidx.recyclerview.widget.LinearLayoutManager$LayoutState.next(LinearLayoutManager.java:2230)
	at androidx.recyclerview.widget.LinearLayoutManager.layoutChunk(LinearLayoutManager.java:1557)
	at androidx.recyclerview.widget.LinearLayoutManager.fill(LinearLayoutManager.java:1517)
	at androidx.recyclerview.widget.LinearLayoutManager.onLayoutChildren(LinearLayoutManager.java:612)
	at androidx.recyclerview.widget.RecyclerView.dispatchLayoutStep2(RecyclerView.java:3924)
	at androidx.recyclerview.widget.RecyclerView.dispatchLayout(RecyclerView.java:3641)
	at androidx.recyclerview.widget.RecyclerView.onLayout(RecyclerView.java:4194)
	at android.view.View.layout(View.java:19659)
	at android.view.ViewGroup.layout(ViewGroup.java:6075)
	at android.widget.RelativeLayout.onLayout(RelativeLayout.java:1083)
	at android.view.View.layout(View.java:19659)
	at android.view.ViewGroup.layout(ViewGroup.java:6075)
	at androidx.viewpager.widget.ViewPager.onLayout(ViewPager.java:1775)
	at android.view.View.layout(View.java:19659)
	at android.view.ViewGroup.layout(ViewGroup.java:6075)
	at android.widget.RelativeLayout.onLayout(RelativeLayout.java:1083)
	at android.view.View.layout(View.java:19659)
	at android.view.ViewGroup.layout(ViewGroup.java:6075)
	at android.widget.FrameLayout.layoutChildren(FrameLayout.java:323)
	at android.widget.FrameLayout.onLayout(FrameLayout.java:261)
	at android.view.View.layout(View.java:19659)
	at android.view.ViewGroup.layout(ViewGroup.java:6075)
	at android.widget.FrameLayout.layoutChildren(FrameLayout.java:323)
	at android.widget.FrameLayout.onLayout(FrameLayout.java:261)
	at android.view.View.layout(View.java:19659)
	at android.view.ViewGroup.layout(ViewGroup.java:6075)
	at androidx.drawerlayout.widget.DrawerLayout.onLayout(DrawerLayout.java:1231)
	at android.view.View.layout(View.java:19659)
	at android.view.ViewGroup.layout(ViewGroup.java:6075)
	at android.widget.FrameLayout.layoutChildren(FrameLayout.java:323)
	at android.widget.FrameLayout.onLayout(FrameLayout.java:261)
	at android.view.View.layout(View.java:19659)
	at android.view.ViewGroup.layout(ViewGroup.java:6075)
	at android.widget.LinearLayout.setChildFrame(LinearLayout.java:1791)
	at android.widget.LinearLayout.layoutVertical(LinearLayout.java:1635)
	at android.widget.LinearLayout.onLayout(LinearLayout.java:1544)
	at android.view.View.layout(View.java:19659)
	at android.view.ViewGroup.layout(ViewGroup.java:6075)
	at android.widget.FrameLayout.layoutChildren(FrameLayout.java:323)
	at android.widget.FrameLayout.onLayout(FrameLayout.java:261)
	at android.view.View.layout(View.java:19659)
	at android.view.ViewGroup.layout(ViewGroup.java:6075)
	at android.widget.LinearLayout.setChildFrame(LinearLayout.java:1791)
	at android.widget.LinearLayout.layoutVertical(LinearLayout.java:1635)
	at android.widget.LinearLayout.onLayout(LinearLayout.java:1544)
	at android.view.View.layout(View.java:19659)
	at android.view.ViewGroup.layout(ViewGroup.java:6075)
	at android.widget.FrameLayout.layoutChildren(FrameLayout.java:323)
	at android.widget.FrameLayout.onLayout(FrameLayout.java:261)
	at com.android.internal.policy.DecorView.onLayout(DecorView.java:761)
	at android.view.View.layout(View.java:19659)
	at android.view.ViewGroup.layout(ViewGroup.java:6075)
	at android.view.ViewRootImpl.performLayout(ViewRootImpl.java:2496)
	at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2212)
	at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1392)
	at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:6752)
	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:911)
	at android.view.Choreographer.doCallbacks(Choreographer.java:723)
	at android.view.Choreographer.doFrame(Choreographer.java:658)
	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:897)
	at android.os.Handler.handleCallback(Handler.java:790)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loop(Looper.java:164)
	at android.app.ActivityThread.main(ActivityThread.java:6494)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:440)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)

```
</p></details>
<hr>

n-xlkt commented:
@vitSkalicky as a build on this feedback, not having insight into the progress of the feed update leads to confusion/obscurity. the first time you start scrolling only to have the line items jump around, you figure you'll wait around longer before scrolling next time. however, it's not clear when it's 'safe' to scroll. PR#2309 avoided this issue with a progress indicator that took over the _What's New_ view, but it's clear this PR is trying to retain the feed while it's updating, and paint as soon as possible. for that reason, i'd suggest something like a sticky-on-scroll, thin progress bar below the tabs. when combined with a 'swipe to refresh', the user will have enough access to information and control to feel comfortable with the feature.

timfbfbfbfb commented:
@Ozyc 
The first and the third exceptions are fixed in the latest version, though the third exception is not introduced by my code changes.
I could not reproduce the second exception, maybe you can try if it is still reproducible in the latest version.

@n-xlkt @cool-student
I added the progress bar in the latest version. Since there are not so many code changes, maybe it is a good idea to just add the progress bar in this PR.

The latest apk is updated in the description, feel free to download and test it.

Mrmusscle commented:
Hi, I had been using your work for days and it was wonderful. Just a small question, my 'live' video keep getting burried for some reason, can you do something about it? 
https://i.imgur.com/V6qJyd4.jpg

Faeb35 commented:
Hi Buddies
Just tested app-a10e6dde4.apk.zip. I really like this implementation, although there are two things I like to mention:
1. The "Whats New" is getting reloaded every time I visit this tab. It's very annoying, because the list is constantly changing while loading. And as I have a lot of channels, that always takes a while... Wouldn't it be possible to store the result and reload/update like every day in the background? My suggestion would be to start a manually reload, when you swipe down (like in Chrome or other apps).
2. I'm having "What's New" configured as my first tab. When I watch a video and come back to the "What's New" list, the list is empty. I'm using Android 5 and the app in german.
![image](https://user-images.githubusercontent.com/9485516/70374947-decd9000-193b-11ea-854d-e58e3d39f470.png)
All the best,
Faeb35

EthanZeigler commented:
This is working flawlessly for me. I think it's stable enough to become an option in prod.

njmdietrich commented:
Love this and have been using it without any issues for the last couple of months now. Last night an issue broke video playback in newspipe and a hotfix has been released. Could you please update the apk here to include this version?

timfbfbfbfb commented:
@njmdietrich The apk file in the description has been updated.

JLammer commented:
@timfbfbfbfb 

Do you have any other plan for this fork, as in add new features o fixes?

Thank you for the sort feature and maintenece!

timfbfbfbfb commented:
@JLammer 
Seems the NewPipe maintainers favor TeamNewPipe/NewPipe#2309 (see the [comment here](https://github.com/TeamNewPipe/NewPipe/pull/2309#issuecomment-578440487)) and probably this PR would not be getting merged, so I guess I would not add features or make big changes to this PR. But since I use this fork myself, I will keep this forked version of APK usable as long as I am still using it.
Thanks for your support anyway :+1:

njmdietrich commented:
I wonder why they prefer the other PR, I think yours is more elegant (from an end user point of view) with the way things load etc, the only thing that's missing from yours to make it perfect in my mind would be a button or pulling down to refresh the list. Props to you and I'm very grateful that you're keeping things updated here :)

JLammer commented:
> the only thing that's missing from yours to make it perfect in my mind would be a button or pulling down to refresh the list.

Indeed! This is the only thing missing <3.

Because I have got subscribed to many channels, fetching new videos every time I go to "what's new" section takes too long. Doing this without consent of the user is really a pain in the butt. Also, If a button to pull down requires a more complex code (hence coding time), providing just a static button would be fine.

Beyond that (although I don't know if it's possible), would be better to fetch and update the "what's new" section only with the newest videos uploaded from the last time you checked until now. In this way, I think you'd update faster and use less bandwidth. I'm pretty sure the app keeps fetching many metadata it already has stored.

PeterHindes commented:
> > the only thing that's missing from yours to make it perfect in my mind would be a button or pulling down to refresh the list.
> 
> Indeed! This is the only thing missing <3.
> 
> Because I have got subscribed to many channels, fetching new videos every time I go to "what's new" section takes too long. Doing this without consent of the user is really a pain in the butt. Also, If a button to pull down requires a more complex code (hence coding time), providing just a static button would be fine.
> 
> Beyond that (although I don't know if it's possible), would be better to fetch and update the "what's new" section only with the newest videos uploaded from the last time you checked until now. In this way, I think you'd update faster and use less bandwidth. I'm pretty sure the app keeps fetching many metadata it already has stored.

This would probably not use less data. If I'm remembering correctly the way to fetch the subscriptions videos is to check the complete video list xml for each channel, and there was no way to only fetch a specific number of entries.

gillhash commented:
@timfbfbfbfb Could you provide an apk with the latest changes?

timfbfbfbfb commented:
> @timfbfbfbfb Could you provide an apk with the latest changes?

Sure. Updated in the description.

wb9688 commented:
Closed in favor of @mauriciocolli's solution.
